### PR TITLE
retry loading saml metadata, log parsing errors

### PR DESCRIPTION
**Pull request by kreek**

### CODE REVIEW

File: lib/saml/settings_service.rb [hunk 1]
```diff
@@ -4,45 +4,59 @@ module SAML
...
+        @saml_settings ||= OneLogin::RubySaml::IdpMetadataParser.new.parse(metadata, settings: settings)
+      rescue StandardError => e
+        Rails.logger.error "SAML::SettingService failed to parse SAML metadata: #{e.message}"
+        raise e
+      end
+
+      private
+
+      def connection
+        @connection ||= Faraday.new(SAML_CONFIG['metadata_url']) do |conn|
```
Comment (omgitsbillryan) [inline lib/saml/settings_service.rb]: The previous code using `Net::Http` was basically a copy+paste from the `ruby-saml` source for reasons that I hope @aub can help me remember.  Is there a reason you switched to Faraday?

File: lib/saml/settings_service.rb [hunk 2]
```diff
@@ -4,45 +4,59 @@ module SAML
...
+          conn.adapter :net_http
+        end
+      end
+
+      def metadata
+        return @metadata unless @metadata.nil?
+        begin
+          try ||= 0
+          response = connection.get
+          raise SAML::InternalServerError, response.status if (500..504).cover? response.status.to_i
```
Comment (omgitsbillryan) [inline lib/saml/settings_service.rb]: Scope creep perhaps, but why not expand this to 400-level response codes also?  Would be nice it we could make use of your `Rails.logger.error` logic instead of the code marching along with a `nil` `response.body` and ultimately spitting some obscure `ruby-saml` exception stack trace to the logs.

File: lib/saml/settings_service.rb (hunk 1 above)
Comment (kreek) [inline lib/saml/settings_service.rb]: @omgitsbillryan using Faraday to standardize responses across the app, it still uses Net::Http under the hood `conn.adapter :net_http`. Faraday does the SSL orchestration for you and makes handling http errors easier.

File: lib/saml/settings_service.rb [hunk 3]
```diff
@@ -4,45 +4,59 @@ module SAML
...
+          @metadata = response.body
+        rescue StandardError => e
+          Rails.logger.error "Failed to load SAML metadata: #{e.message}: try #{try + 1} of #{METADATA_RETRIES}"
+          sleep try * 0.25
```
Comment (kreek) [inline lib/saml/settings_service.rb]: 👍 good catch, sleep backoff should be inside the retry conditional

File: lib/saml/settings_service.rb [hunk 4]
```diff
@@ -1,48 +1,67 @@
...
-      settings.private_key  = SAML_CONFIG['key']
-      settings.issuer       = SAML_CONFIG['issuer']
-      settings.assertion_consumer_service_url = SAML_CONFIG['callback_url']
+      def metadata
+        attempt ||= 0
+        response = connection.get
+        raise SAML::InternalServerError, response.status if (400..504).cover? response.status.to_i
+        response.body
+      rescue StandardError => e
+        Rails.logger.error "Failed to load SAML metadata: #{e.message}: try #{attempt + 1} of #{METADATA_RETRIES}"
```
Comment (omgitsbillryan) [inline lib/saml/settings_service.rb]: `Rails.logger.error` could be harsh for attempts 1 & 2.    The whole premise of this PR is that a call or two to get metadata _might_ fail and we recover by trying again.  I wouldn't want an `ERROR` level message to create a red herring in the logs in the event that the first call failed but we successfully recovered in the 2nd call.  What about this instead:

```ruby
msg = "Failed to load SAML metadata: #{e.message}: try #{attempt + 1} of #{METADATA_RETRIES}"
(attempt >= METADATA_RETRIES) ? Rails.logger.error msg : Rails.logger.warn msg
```
